### PR TITLE
Implemented integrity check (md5 hash) after dataset download

### PR DESCRIPTION
**Pull request by le1nux**
With regard to issue #1415, I implemented md5 hash check after downloading a compressed dataset file and **before** extracting it. 
See [here](https://docs.python.org/3/library/tarfile.html#tarfile.TarFile.extractall) for security risks, when extracting archives in Python.

I also added missing md5 hashes for the datasets  MNIST, FashionMNIST, KMNIST, EMNIST and QMNIST.

### CODE REVIEW

**Comment by pmeier** [comment]
Couldn't we add this check at the end of `download_url()`?

https://github.com/pytorch/vision/blob/ed5b2dc3a5e7411d8b40cc7e526e151983e99cf9/torchvision/datasets/utils.py#L59-L60

With this, if we have an MD5, your change is automatically included in every dataset.

File: torchvision/datasets/utils.py [hunk 1]
```diff
@@ -95,7 +95,10 @@ def download_url(url, root, filename=None, md5=None):
                 )
             else:
                 raise e
-
+        # check integrity of downloaded file
+        if not check_integrity(fpath, md5):
+            print("Error downloading file: Dataset not found in " + fpath + " or corrupted.")
```
Comment (pmeier) [inline torchvision/datasets/utils.py]: 1. This is not only used to download datasets. Maybe change to wording from "Dataset" to "file"?
2. Is there a possibility that the file is not present after the download? If not, I would remove the "not found" part

Comment (le1nux) [reply torchvision/datasets/utils.py]: 1. Yes, totally agree. 
2. `check_integrity` checks first whether a file is present or not. So I included that information in  the log message. However, I'm not sure if [urlretrieve](https://docs.python.org/3/library/urllib.request.html#urllib.request.urlretrieve) would always create a file. It also seems to be a deprecated function.

File: torchvision/datasets/utils.py [hunk 2]
```diff
@@ -95,7 +95,10 @@ def download_url(url, root, filename=None, md5=None):
...
+            raise RuntimeError("Dataset not found or corrupted.")
```
Comment (fmassa) [inline torchvision/datasets/utils.py]: Is there a need to print a message and raise a `RuntimeError`? Can't you just raise the error?

File: torchvision/datasets/mnist.py [hunk 3]
```diff
@@ -172,11 +174,11 @@ class FashionMNIST(MNIST):
         target_transform (callable, optional): A function/transform that takes in the
             target and transforms it.
     """
-    urls = [
-        'http://fashion-mnist.s3-website.eu-central-1.amazonaws.com/train-images-idx3-ubyte.gz',
-        'http://fashion-mnist.s3-website.eu-central-1.amazonaws.com/train-labels-idx1-ubyte.gz',
-        'http://fashion-mnist.s3-website.eu-central-1.amazonaws.com/t10k-images-idx3-ubyte.gz',
-        'http://fashion-mnist.s3-website.eu-central-1.amazonaws.com/t10k-labels-idx1-ubyte.gz',
+    resources = [
+        ("http://fashion-mnist.s3-website.eu-central-1.amazonaws.com/train-images-idx3-ubyte.gz", "8d4fb7e6c68d591d4c3dfef9ec88bf0d"),
```
Comment (fmassa) [inline torchvision/datasets/mnist.py]: Those lines are too long and make lint fail

Comment (le1nux) [reply torchvision/datasets/mnist.py]: Done.

**Comment by fmassa** [comment]
@le1nux can you fix lint and rebase your PR on top of master? This is almost good to be merged.

**Comment by le1nux** [comment]
Sorry, I was covered in work the past days. Finally I got to rebase the PR, today. Any remarks?

**Pull request merged**